components/tidb_query/src/codec/row/v2/encoder.rs

```diff
@@ -0,0 +1,296 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+
+use crate::codec::{
+    mysql::{decimal::DecimalEncoder, json::JsonEncoder},
+    Datum, Error, Result,
+};
+use codec::prelude::*;
+
+const OFFSET_SIZE: usize = 2;
+const LARGE_OFFSET_SIZE: usize = 4;
+const ID_SIZE: usize = 1;
+const LARGE_ID_SIZE: usize = 4;
+
+// the row format is:
+// | version | flag | number_of_not_null_values | number_of_null_values | not_null_value_ids | null_value_ids | value_offsets | values
+// short spec of each field, check https://github.com/pingcap/tidb/blob/master/docs/design/2018-07-19-row-format.md for details
+// version: 1 byte
+// flag: 1 byte, when there's id greater than 255, value is 1, otherwise 0
+// number of not null values: 2 bytes
+// number of null values: 2 bytes
+// column ids:  ids of not null values + ids of null values, when flag == 1 (large), id is 4 bytes, otherwise 1 byte
+// not null values offset: when large, offset is 4 bytes, otherwise 2 bytes
+pub fn encode(datums: Vec<Datum>, ids: &[i64]) -> Result<Vec<u8>> {
+    let row = Row::new_row(datums, ids);
+    let mut wtr = Vec::with_capacity(row.approx_size());
+    wtr.encode_row(row)?;
+    Ok(wtr)
+}
+
+trait DatumEncoder: NumberEncoder {
+    // encode_json & encode_decimal does not use codec::BufferWriter yet
+    fn encode_datum(&mut self, datum: Datum) -> Result<()>;
+
+    #[allow(clippy::match_overlapping_arm)]
+    #[inline]
+    fn encode_i64(&mut self, v: i64) -> codec::Result<()> {
+        match v {
+            -128..=127 => self.write_u8(v as i8 as u8),
```

Comment:
It's better to use constants defined in the standard library.
```suggestion
            i8::MIN..=i8::MAX => self.write_u8(v as i8 as u8),
```